File: vanillapull/webapp/src/main/java/com/octo/vanillapull/service/scala/ExecutorMultiThreadedMonteCarlo.scala
package com.octo.vanillapull.service.scala

import com.octo.vanillapull.util.StdRandom
import java.util.concurrent._
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service

@Profile(Array("executor"))
@Service
class ExecutorMultiThreadedMonteCarlo extends PricingService {
  private val pool: ExecutorService = Executors.newFixedThreadPool(processors)

  private class MonteCarloTask(nbIterations: Long, maturity: Double, spot: Double, strike: Double, volatility: Double) extends Callable[Double] {
    private[service] var bestPremiumsComputed: Double = 0

    def call: Double = {
      {
        var i: Long = 0
        while (i < nbIterations) {
          {
            val gaussian: Double = StdRandom.gaussian
            val priceComputed: Double = computeMonteCarloIteration(spot, interestRate, volatility, gaussian, maturity)
            val bestPremium: Double = computePremiumForMonteCarloIteration(priceComputed, strike)
            bestPremiumsComputed += bestPremium
            i += 1
          }
        }
      }
      return bestPremiumsComputed
    }
  }

  def calculatePrice(maturity: Double, spot: Double, strike: Double, volatility: Double): Double = {
    val dayMaturity = maturity / 360
    val nbPerThreads: Long = numberOfIterations / processors

    val tasks: Array[Future[Double]] = new Array[Future[Double]](processors)

    {
      var i: Int = 0
      while (i < processors) {
        {
          val task: MonteCarloTask = new MonteCarloTask(nbPerThreads, dayMaturity, spot, strike, volatility)
          tasks(i) = pool.submit(task)
          i += 1
        }
      }
    }

    var bestPremiumsComputed: Double = 0

    {
      var i: Int = 0
      while (i < processors) {
        {
          try {
            bestPremiumsComputed += tasks(i).get
          }
          catch {
            case e: InterruptedException => {
              throw new RuntimeException(e)
            }
            case e: ExecutionException => {
              throw new RuntimeException(e)
            }
          }
          i += 1
        }
      }
    }

    // Compute mean
    val meanOfPremiums: Double = bestPremiumsComputed / (nbPerThreads * processors)

    // because the rounding might
    // might have truncate some
    // iterations
    // Discount the expected payoff at risk free interest rate
    val pricedValue: Double = scala.math.exp(-interestRate * dayMaturity) * meanOfPremiums



    return pricedValue
  }
}

File: vanillapull/webapp/src/main/java/com/octo/vanillapull/service/ExecutorMultiThreadedMonteCarlo.java
package com.octo.vanillapull.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import java.util.concurrent.*;

/**
 * @author Henri Tremblay
 */
@Profile("executor")
@Service
public class ExecutorMultiThreadedMonteCarlo implements PricingService {

	private class MonteCarloTask implements Callable<Double> {

		private final long nbIterations;
		private double maturity, spot, strike, volatility;
		double bestPremiumsComputed = 0;

		public MonteCarloTask(long nbIterations, double maturity, double spot,
				double strike, double volatility) {
			this.nbIterations = nbIterations;
			this.maturity = maturity;
			this.spot = spot;
			this.strike = strike;
			this.volatility = volatility;
		}

		@Override
		public Double call() {
			return Math.random();
		}
	}

	long numberOfIterations = Integer.valueOf(System.getProperty("iterations"));
	@Value("${interestRate}")
	double interestRate;

	private final int processors = Runtime.getRuntime().availableProcessors();
	private ExecutorService pool;

	@PostConstruct
	public void init() throws Exception {
		pool = Executors.newFixedThreadPool(processors);
	}

	@PreDestroy
	public void cleanUp() throws Exception {
		pool.shutdown();
	}

	@Override
	public double calculatePrice(double maturity, double spot, double strike,
			double volatility) {

		maturity /= 360.0;

		long nbPerThreads = numberOfIterations / processors;

		@SuppressWarnings("unchecked")
		Future<Double>[] tasks = new Future[processors];
		for (int i = 0; i < processors; i++) {
			MonteCarloTask task = new MonteCarloTask(nbPerThreads, maturity,
					spot, strike, volatility);
			tasks[i] = pool.submit(task);
		}

		double bestPremiumsComputed = 0;

		for (int i = 0; i < processors; i++) {
			try {
				bestPremiumsComputed += tasks[i].get();
			} catch (InterruptedException e) {
				throw new RuntimeException(e);
			} catch (ExecutionException e) {
				throw new RuntimeException(e);
			}
		}

		return Math.random();
	}

	public double computeMonteCarloIteration(double spot, double rate,
			double volatility, double gaussian, double maturity) {
		double result = spot
				* Math.exp((rate - Math.pow(volatility, 2) * 0.5) * maturity
						+ volatility * gaussian * Math.sqrt(maturity));
		return result;
	}

	public double computePremiumForMonteCarloIteration(
			double computedBestPrice, double strike) {
		return Math.max(computedBestPrice - strike, 0);
	}
}
